cc/position.cc

```diff
@@ -424,6 +429,264 @@ float Position::CalculateScore(float komi) {
+      Coord c(row, col);
+      if (stones_[c].color() != color || !board_visitor_->Visit(c)) {
+        continue;
+      }
+
+      auto g = absl::make_unique<BensonGroup>();
+      g->first_stone = c;
+      while (!board_visitor_->Done()) {
+        c = board_visitor_->Next();
+        groups[c] = g.get();
+
+        for (auto nc : kNeighborCoords[c]) {
+          auto ns = stones_[nc];
+          if (ns.empty()) {
+            g->liberties.insert(nc);
+          } else if (ns.color() == color) {
+            board_visitor_->Visit(nc);
+          }
+        }
+      }
+      group_pool.push_back(std::move(g));
+    }
+  }
+
+  // Build the set of all regions.
+  board_visitor_->Begin();
+  for (int row = 0; row < kN; ++row) {
+    for (int col = 0; col < kN; ++col) {
+      Coord c(row, col);
+      if (stones_[c].color() == color || !board_visitor_->Visit(c)) {
+        continue;
+      }
+
+      auto r = absl::make_unique<BensonRegion>();
+      r->first_stone = c;
+      while (!board_visitor_->Done()) {
+        c = board_visitor_->Next();
+
+        regions[c] = r.get();
+        if (stones_[c].empty()) {
+          r->empty_points.insert(c);
+        } else {
+          r->other_color_points.insert(c);
+        }
+
+        for (auto nc : kNeighborCoords[c]) {
+          auto ns = stones_[nc];
+          if (ns.color() != color) {
+            board_visitor_->Visit(nc);
+          } else {
+            r->enclosing_groups.insert(groups[nc]);
+          }
+        }
+      }
+
+      // Find the vital groups for this region.
+      for (auto* g : r->enclosing_groups) {
+        bool is_vital =
+            std::includes(g->liberties.begin(), g->liberties.end(),
+                          r->empty_points.begin(), r->empty_points.end());
+        if (is_vital) {
+          r->vital_for.insert(g);
+          g->num_vital_regions += 1;
+        }
+      }
+
+      region_pool.push_back(std::move(r));
+    }
+  }
+
+  // Print some debug information.
+  for (auto& r : region_pool) {
+    MG_LOG(INFO) << "REGION: " << r->first_stone.ToGtp();
+    for (auto* g : r->vital_for) {
+      MG_LOG(INFO) << "  VITAL: " << g->first_stone.ToGtp();
+    }
+    for (auto* g : r->enclosing_groups) {
+      MG_LOG(INFO) << "  NEIGHBOR: " << g->first_stone.ToGtp();
+    }
+  }
+  for (auto& g : group_pool) {
+    MG_LOG(INFO) << "GROUP: " << g->first_stone.ToGtp() << " "
+                 << g->num_vital_regions;
+    for (auto c : g->liberties) {
+      MG_LOG(INFO) << "  LIBERTY: " << c.ToGtp();
+    }
+  }
+
+  // Run Benson's algorithm.
+  for (;;) {
+    // List of groups removed this iteration.
+    std::vector<std::unique_ptr<BensonGroup>> removed_groups;
+
+    // Iterate over remaining groups.
+    for (size_t i = 0; i < group_pool.size();) {
+      if (group_pool[i]->num_vital_regions < 2) {
+        // This group has fewer than two vital regions, remove it.
+        removed_groups.push_back(std::move(group_pool[i]));
```

Comment:
I guess I don't understand what std::move is doing here or why you have to manually fix up the back side of the group_pool vector. Why aren't these three lines just `removed_groups.push_back(group_pool.pop_back())`?

Reply:
We're removing element `i`, not the last element.

Since we don't care about the order of elements in `group_pool`, we can remove element `i` without having to shuffle all the later elements over by swapping `i` and the last element, then popping the last element (which used to be `i`) off.